### ISSUE TITLE
Desaturate Artificating

### ISSUE THREAD

**Issue by andrewjhaman** [alecjacobson/computer-graphics-raster-images#60]
There seems to be some artificating in the example image given for desaturate.cpp. 

If you look closely at the dogs closest ear in desaturated.png there is a region where the saturation is all 0 and there are blocky chunks that do not change smoothly in and out
of the region.

When I tried desaturating the original image using photoshop I was not able to replicate these same artifacts.

Is this the desired result of our implementation as well?

**Comment by lzm0** [comment]
I also tried my implementation of desaturation, no artifacts found compared to the example. But I think that the example images are converted to png, which could be the cause.
This is my result, which is also converted to png, with no artifacts:
![desaturated](https://user-images.githubusercontent.com/7163127/104757851-869cca80-5798-11eb-9c05-a36bb3025fec.png)

**Comment by andrewjhaman** [comment]
I was able to replicate the same pattern of artifacts in my code however not the same values, instead mine are filled with greenish values.
![desaturated](https://user-images.githubusercontent.com/49106639/104759313-f53c4280-571c-11eb-9ea8-eedfedfdd731.png)

**Comment by lzm0** [comment]
> I was able to replicate the same pattern of artifacts in my code however not the same values, instead mine are filled with greenish values.
> ![desaturated](https://user-images.githubusercontent.com/49106639/104759313-f53c4280-571c-11eb-9ea8-eedfedfdd731.png)

Actually I had very similar pattern in my first version. It could be rgb_to_hsv that is not well implemented. (notice that the value of h you calculated can be out of range and you have to shift it to [0,360) )

**Comment by andrewjhaman** [comment]
Indeed, I am able to achieve what I believe to be the correct result without artifacts, it is just not the same as the example image given. It seems like they may be clamping their hue value instead correctly wrapping it.

**Comment by rozariot** [comment]
I also seem to be getting a desaturated image that is different from their example. I think it is the same as both of your images.

Can someone please let us know if we are doing this incorrectly?

![desaturated](https://user-images.githubusercontent.com/39061301/104820683-9225db00-5804-11eb-89e1-5b48f9ed76e3.png)
